### PR TITLE
Start work providers feature

### PR DESCRIPTION
**Pull request by dmitry-shibanov**
**Description:**
Draft pull request to check current status of v2 setup-java.

**Related issue:**
Add link to the related issue.

**Check list:**
- [ ] Mark if documentation changes are required.
- [ ] Mark if tests were added or updated to cover the changes.

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -21,6 +21,12 @@ inputs:
     description: 'Path to where the compressed JDK is located. The path could
        be in your source repository or a local path on the agent.'
     required: false
+  provider:
```
Comment (maxim-lobanov) [inline action.yml:24]: Please rename this property according to ADR.

File: action.yml [hunk 2]
```diff
@@ -21,6 +21,12 @@ inputs:
...
+    description: 'Provider of java'
+    required: false
+  feature:
```
Comment (maxim-lobanov) [inline action.yml:27]: Let's get rid of this for now

File: src/installer-updated.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import * as core from '@actions/core';
```
Comment (maxim-lobanov) [inline src/installer-updated.ts:1]: This file has very strange name: `installer-updated.ts`

File: src/providers/java-factory.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import ZuluProvider from "./zulu-provider";
+
+export enum JavaProviders {
+    AdopOpenJdk = "adopOpenJdk",
+    Zulu = "zulu"
+}
+
+export class JavaFactory {
+
+    constructor(private http: httpm.HttpClient, private version: string, private arch: string, private javaPackage: string = "jdk", private features?: string) { };
```
Comment (maxim-lobanov) [inline src/providers/java-factory.ts:13]: May be we can put this code and factory to `installer-updated.ts`

File: src/setup-java.ts [hunk 5]
```diff
@@ -1,60 +1,69 @@
 import * as core from '@actions/core';
-import * as installer from './installer';
+import * as installer from './installer-updated';
 import * as auth from './auth';
 import * as gpg from './gpg';
-import * as constants from './constants';
+import {
```
Comment (maxim-lobanov) [inline src/setup-java.ts:5]: I think the previous approach with `import * as constants from './constants';` is more readable

File: src/distributors/adoptopenjdk-installer.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        toolPath = await tc.cacheDir(archivePath, `Java_${this.Distributor}_${this.javaPackage}`, javaRelease.resolvedVersion, this.arch);
+
+        if(process.platform === 'darwin') {
+            toolPath = path.join(toolPath, macOSJavaContentDir);
+        }
+
+        return { javaPath: toolPath, javaVersion: javaRelease.resolvedVersion };
+    }
+
+    protected get Distributor(): string {
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:62]: I think it was better when we just pass distributor name as string in contructor as you have implemented initially. Sorry for confusing. Let's use your approach

File: src/distributors/zulu-installer.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        this.arch = this.arch === 'x64' ? 'x86' : this.arch;
+    }
+
+    protected async downloadTool(javaRelease: IJavaRelease): Promise<IJavaInfo> {
+        let toolPath: string;
+        let downloadDir: string;
+
+        const javaPath = await tc.downloadTool(javaRelease.link);
+        
+        core.info(`Ectracting ${this.Distributor} java version ${javaRelease.resolvedVersion}`);
```
Comment (maxim-lobanov) [inline src/distributors/zulu-installer.ts:27]: typo in message `Ectracting`

File: src/distributors/zulu-installer.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+        const archiveName = fs.readdirSync(downloadDir)[0];
+        const archivePath = path.join(downloadDir, archiveName);
+        toolPath = await tc.cacheDir(archivePath, `Java_${this.Distributor}_${this.javaPackage}`, javaRelease.resolvedVersion, this.arch);
+
+        return { javaPath: toolPath, javaVersion: javaRelease.resolvedVersion };
+    }
+
+    protected async resolveVersion(range: semver.Range): Promise<IJavaRelease> {
+        const resolvedVersion = await this.getAvailableVersion(range);
```
Comment (maxim-lobanov) [inline src/distributors/zulu-installer.ts:42]: splitting by functions looks a bit strange for me.

File: src/setup-java.ts [hunk 9]
```diff
@@ -1,63 +1,72 @@
...
-    }
-
+    const javaDistributor = core.getInput('distribution');
     const javaPackage = core.getInput(constants.INPUT_JAVA_PACKAGE, {
       required: true
     });
     const jdkFile = core.getInput(constants.INPUT_JDK_FILE, {required: false});
 
-    await installer.getJava(version, arch, jdkFile, javaPackage);
+    const initOptions: JavaInitOptions = {
```
Comment (maxim-lobanov) [inline src/setup-java.ts:21]: May be `JavaInstallerOptions` or `JavaInstallerInputs` as a name for interface?

Comment (dmitry-shibanov) [reply src/setup-java.ts]: Done.

File: src/distributors/zulu-installer.ts (hunk 7 above)
Comment (dmitry-shibanov) [reply src/distributors/zulu-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 6 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+        return majorVersion;
+    }
+
+    protected async downloadTool(javaRelease: IJavaRelease): Promise<IJavaInfo> {
+        let toolPath: string;
+        let downloadDir: string;
+
+        core.info(`Downloading ${this.Distributor}, java version ${javaRelease.resolvedVersion}`);
```
[the remark this answers is not in the document]
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        
+        if(IS_WINDOWS) {
+            downloadDir = await tc.extractZip(javaPath);
+        } else {
+            downloadDir = await tc.extractTar(javaPath);
+        }
+
+        const archiveName = fs.readdirSync(downloadDir)[0];
+        const archivePath = path.join(downloadDir, archiveName);
+        toolPath = await tc.cacheDir(archivePath, `Java_${this.Distributor}_${this.javaPackage}`, javaRelease.resolvedVersion, this.arch);
```
[the remark this answers is not in the document]
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+import { IS_WINDOWS, PLATFORM, IS_MACOS, macOSJavaContentDir } from "../util";
+import { IJavaInfo, IJavaRelease, JavaBase, JavaInitOptions } from "./base-installer";
+import { IRelease, IReleaseVersion } from './adoptopenjdk-models'
+
+export class AdoptOpenJdkDistributor extends JavaBase {
+    private platform: string;
+    
+    constructor(initOptions: JavaInitOptions) {
+        super(initOptions);
+        this.platform = IS_MACOS ? 'mac' : PLATFORM;
```
[the remark this answers is not in the document]
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/auth.ts [hunk 13]
```diff
@@ -13,23 +14,22 @@ export async function configAuthentication(
...
+     password=\$${password},
+     and gpg-passphrase=${gpgPassphrase ? '$' + gpgPassphrase : null}`
   );
   // when an alternate m2 location is specified use only that location (no .m2 directory)
   // otherwise use the home/.m2/ path
   const settingsDirectory: string = path.join(
-    core.getInput(constants.INPUT_SETTINGS_PATH) || os.homedir(),
-    core.getInput(constants.INPUT_SETTINGS_PATH) ? '' : M2_DIR
+    core.getInput(INPUT_SETTINGS_PATH) || os.homedir(),
+    core.getInput(INPUT_SETTINGS_PATH) || M2_DIR
```
[the remark this answers is not in the document]
Comment (dmitry-shibanov) [reply src/auth.ts]: Done.

File: src/auth.ts [hunk 14]
```diff
@@ -3,8 +3,9 @@ import * as os from 'os';
 import * as path from 'path';
 import * as core from '@actions/core';
 import * as io from '@actions/io';
+
 import {create as xmlCreate} from 'xmlbuilder2';
-import * as constants from './constants';
+import {INPUT_SETTINGS_PATH} from './constants';
```
Comment (maxim-lobanov) [inline src/auth.ts:8]: revert constants

File: src/distributors/adoptopenjdk-installer.ts [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+import { IRelease, IReleaseVersion } from './adoptopenjdk-models'
+
+export class AdoptOpenJdkDistributor extends JavaBase {
+    
+    constructor(initOptions: JavaInstallerOptions) {
+        super("AdoptOpenJDK", initOptions);
+    }
+
+    private async getAvailableMajor(range: semver.Range) {
+        const urlReleaseVersion = "https://api.adoptopenjdk.net/v3/info/available_releases"
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:19]: consider adding `lts` input

File: src/distributors/adoptopenjdk-installer.ts [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    protected get javaRootName(): string {
+        return `Java_${this.distributor}_${this.javaPackage}`;
+    }
+
+    protected async resolveVersion(range: semver.Range): Promise<IJavaRelease> {
+        const platform = IS_MACOS ? 'mac' : PLATFORM;
+
+        const majorVersion = await this.getAvailableMajor(range);
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:67]: `resolvedMajorVersion`

File: src/distributors/adoptopenjdk-installer.ts (hunk 15 above)
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:19]: `availableVersionsReleaseUrl`

File: src/distributors/adoptopenjdk-installer.ts [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+export class AdoptOpenJdkDistributor extends JavaBase {
+    
+    constructor(initOptions: JavaInstallerOptions) {
+        super("AdoptOpenJDK", initOptions);
+    }
+
+    private async getAvailableMajor(range: semver.Range) {
+        const urlReleaseVersion = "https://api.adoptopenjdk.net/v3/info/available_releases"
+        const javaVersionAvailable = (await this.http.getJson<IReleaseVersion>(urlReleaseVersion)).result;
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:20]: `availableMajorVersions`

File: src/distributors/adoptopenjdk-installer.ts [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+
+        if (!javaVersionAvailable) {
+            throw new Error(`No versions were found for ${this.distributor}`)
+        }
+
+        const javaSemVer = javaVersionAvailable.available_releases.map(item => semver.coerce(item)!)!;
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:26]: coercedAvailableVersions

File: src/distributors/adoptopenjdk-installer.ts [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+        const majorVersion = semver.maxSatisfying(javaSemVer, range)?.major;
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:27]: resolvedMajorVersion

File: src/distributors/adoptopenjdk-installer.ts [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+    protected get javaRootName(): string {
+        return `Java_${this.distributor}_${this.javaPackage}`;
+    }
+
+    protected async resolveVersion(range: semver.Range): Promise<IJavaRelease> {
+        const platform = IS_MACOS ? 'mac' : PLATFORM;
+
+        const majorVersion = await this.getAvailableMajor(range);
+        const releasesUrl = `https://api.adoptopenjdk.net/v3/assets/feature_releases/${majorVersion}/ga?heap_size=normal&image_type=${this.javaPackage}&page=0&page_size=1000&project=jdk&sort_method=DEFAULT&sort_order=DESC&vendor=adoptopenjdk&jvm_impl=hotspot&architecture=${this.arch}&os=${platform}`;
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:68]: availableVersionsUrl

File: src/distributors/adoptopenjdk-installer.ts [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+        const javaVersionReleases = ( await this.http.getJson<IRelease[]>(releasesUrl)).result;
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:69]: availableVersionsList

File: src/distributors/adoptopenjdk-installer.ts [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+        const fullVersion = javaVersionReleases?.find(item => semver.satisfies(item.version_data.semver, range));
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:70]: resolvedFullVersion

File: src/distributors/adoptopenjdk-installer.ts [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+        toolPath = await tc.cacheDir(archivePath, `Java_${this.distributor}_${this.javaPackage}`, javaRelease.resolvedVersion, this.arch);
+
+        if(process.platform === 'darwin') {
+            toolPath = path.join(toolPath, macOSJavaContentDir);
+        }
+
+        return { javaPath: toolPath, javaVersion: javaRelease.resolvedVersion };
+    }
+
+    protected get javaRootName(): string {
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:60]: toolcacheFolderName

File: src/distributors/adoptopenjdk-installer.ts [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+
+        return majorVersion;
+    }
+
+    protected async downloadTool(javaRelease: IJavaRelease): Promise<IJavaInfo> {
+        let toolPath: string;
+        let downloadDir: string;
+
+        core.info(`Downloading ${this.distributor}, java version ${javaRelease.resolvedVersion}`);
+        const javaPath = await tc.downloadTool(javaRelease.link);
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:41]: javaArchivePath

File: src/distributors/adoptopenjdk-installer.ts [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+        
+        if(IS_WINDOWS) {
+            downloadDir = await tc.extractZip(javaPath);
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:44]: extractedJavaPath

File: src/distributors/adoptopenjdk-installer.ts (hunk 23 above)
Comment (maxim-lobanov) [reply src/distributors/adoptopenjdk-installer.ts:60]: please moved this function to base class

File: src/distributors/base-installer.ts [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+import * as httpm from '@actions/http-client';
+import {  getVersionFromToolcachePath, parseLocalVersions } from '../util';
+
+export interface JavaInstallerOptions {
+    version: string;
+    arch: string;
+    javaPackage: string;
+}
+
+export interface IJavaInfo {
```
Comment (maxim-lobanov) [inline src/distributors/base-installer.ts:14]: `JavaInstallerResults`

File: src/distributors/base-installer.ts [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+    javaVersion: string;
+    javaPath: string;
+}
+
+export interface IJavaRelease {
```
Comment (maxim-lobanov) [inline src/distributors/base-installer.ts:19]: JavaDownloadRelease

File: src/distributors/base-installer.ts [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+    protected abstract downloadTool(javaRelease: IJavaRelease): Promise<IJavaInfo>;
+    protected abstract resolveVersion(range: semver.Range): Promise<IJavaRelease>;
+    protected abstract get javaRootName(): string;
+
+    public async getJava(): Promise<IJavaInfo> {
+        const range = new semver.Range(this.version);
+        let foundJava = this.findInToolcache(range);
+        if (!foundJava) {
+            // try to find Java in default system locations outside tool-cache
+            foundJava = this.findInKnownLocations(range)
```
Comment (maxim-lobanov) [inline src/distributors/base-installer.ts:48]: Let's remove findInKnownLocations and related logic and move Java to toolcache on images

File: src/distributors/base-installer.ts [hunk 29]
```diff
@@ -0,0 +1,144 @@
...
+            break;
+            default: knownLocation = '/usr/lib/jvm'
+        }
+
+        const localVersions = parseLocalVersions(knownLocation, this.distributor);
+        return localVersions.find(localVersion => semver.satisfies(localVersion.javaVersion, version)) ?? null;
+    }
+
+    protected setJavaDefault(toolPath: string) {
+        const extendedJavaHome = `JAVA_HOME_${this.version}_${this.arch}`
```
Comment (maxim-lobanov) [inline src/distributors/base-installer.ts:93]: let's remove this varialbe

File: src/distributors/base-installer.ts [hunk 30]
```diff
@@ -0,0 +1,144 @@
...
+        switch(process.platform) {
+            case "win32": return path.normalize('C:/Program Files/Java');
+            case "darwin": return '/Library/Java/JavaVirtualMachines'
+            default: return '/usr/lib/jvm'
+        }
+    }
+
+    // this function validates and parse java version to its normal semver notation
+    protected normalizeVersion(version: string): string {
+        if (version.slice(0, 2) === '1.') {
```
Comment (maxim-lobanov) [inline src/distributors/base-installer.ts:113]: use `startsWith`

Comment (dmitry-shibanov) [reply src/distributors/base-installer.ts]: Done.

File: src/distributors/base-installer.ts (hunk 29 above)
Comment (dmitry-shibanov) [reply src/distributors/base-installer.ts]: Done.

File: src/distributors/base-installer.ts (hunk 28 above)
Comment (dmitry-shibanov) [reply src/distributors/base-installer.ts]: Done.

File: src/distributors/base-installer.ts (hunk 26 above)
Comment (dmitry-shibanov) [reply src/distributors/base-installer.ts]: Done.

File: src/distributors/base-installer.ts (hunk 27 above)
Comment (dmitry-shibanov) [reply src/distributors/base-installer.ts:19]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 23 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 24 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 25 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 20 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 21 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 22 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 18 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 19 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 17 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 15 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 16 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/auth.ts (hunk 14 above)
Comment (dmitry-shibanov) [reply src/auth.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts [hunk 31]
```diff
@@ -0,0 +1,79 @@
...
+
+import fs from 'fs';
+import path from 'path';
+import semver from 'semver';
+
+import { IS_WINDOWS, PLATFORM, IS_MACOS, macOSJavaContentDir } from "../util";
+import { JavaInstallerResults, IJavaRelease, JavaBase, JavaInstallerOptions } from "./base-installer";
+import { IRelease, IReleaseVersion } from './adoptopenjdk-models'
+
+export class AdoptOpenJdkDistributor extends JavaBase {
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:12]: nitpick: `AdoptOpenJDKDistributor`

File: src/distributors/adoptopenjdk-installer.ts [hunk 32]
```diff
@@ -0,0 +1,79 @@
...
+
+    private async getAvailableMajor(range: semver.Range) {
+        const availableMajorVersionsUrl = "https://api.adoptopenjdk.net/v3/info/available_releases"
+        const availableMajorVersions = (await this.http.getJson<IReleaseVersion>(availableMajorVersionsUrl)).result;
+
+        if (!availableMajorVersions) {
+            throw new Error(`No versions were found for ${this.distributor}`)
+        }
+
+        const coercedAvailableVersions = availableMajorVersions.available_releases.map(item => semver.coerce(item)!)!;
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:26]: `.map(item => semver.coerce(item)!)!;` looks a bit strange.
May be `.map(item => semver.coerce(item)).filter(item => Boolean(item))`

File: src/distributors/adoptopenjdk-installer.ts [hunk 33]
```diff
@@ -0,0 +1,79 @@
...
+            extractedJavaPath = await tc.extractZip(javaArchivePath);
+        } else {
+            extractedJavaPath = await tc.extractTar(javaArchivePath);
+        }
+
+        const archiveName = fs.readdirSync(extractedJavaPath)[0];
+        const archivePath = path.join(extractedJavaPath, archiveName);
+        toolPath = await tc.cacheDir(archivePath, `Java_${this.distributor}_${this.javaPackage}`, javaRelease.resolvedVersion, this.arch);
+
+        if(process.platform === 'darwin') {
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:53]: nitpick: style around `if (`

File: src/distributors/adoptopenjdk-installer.ts [hunk 34]
```diff
@@ -0,0 +1,79 @@
...
+        const resolvedMajorVersion = await this.getAvailableMajor(range);
+        const availableVersionsUrl = `https://api.adoptopenjdk.net/v3/assets/feature_releases/${resolvedMajorVersion}/ga?heap_size=normal&image_type=${this.javaPackage}&page=0&page_size=1000&project=jdk&sort_method=DEFAULT&sort_order=DESC&vendor=adoptopenjdk&jvm_impl=hotspot&architecture=${this.arch}&os=${platform}`;
+        const availableVersionsList = ( await this.http.getJson<IRelease[]>(availableVersionsUrl)).result;
+        const resolvedFullVersion = availableVersionsList?.find(item => semver.satisfies(item.version_data.semver, range));
+
+        if(!resolvedFullVersion) {
+            throw new Error(`Could not find satisfied version in ${availableVersionsList}`);
+        }
+
+        const javaRelease: IJavaRelease = {
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:72]: Can we just return it?

File: src/distributors/base-installer.ts [hunk 35]
```diff
@@ -0,0 +1,121 @@
...
+    protected http: httpm.HttpClient;
+    protected version: string;
+    protected arch: string;
+    protected javaPackage: string;
+    constructor(protected distributor: string, initOptions: JavaInstallerOptions) {
+        this.http = new httpm.HttpClient('setup-java', undefined, {
+            allowRetries: true,
+            maxRetries: 3
+          });
+          this.version = this.normalizeVersion(initOptions.version);
```
Comment (maxim-lobanov) [inline src/distributors/base-installer.ts:34]: indent looks incorrect

File: src/distributors/base-installer.ts [hunk 36]
```diff
@@ -0,0 +1,121 @@
...
+          this.arch = initOptions.arch;
+          this.javaPackage  = initOptions.javaPackage;
+    }
+
+    protected abstract downloadTool(javaRelease: IJavaRelease): Promise<JavaInstallerResults>;
+    protected abstract resolveVersion(range: semver.Range): Promise<IJavaRelease>;
+
+    public async getJava(): Promise<JavaInstallerResults> {
```
Comment (maxim-lobanov) [inline src/distributors/base-installer.ts:42]: may be `setupJava()`

File: src/distributors/base-installer.ts [hunk 37]
```diff
@@ -0,0 +1,121 @@
...
+        const range = new semver.Range(this.version);
```
Comment (maxim-lobanov) [inline src/distributors/base-installer.ts:43]: may be we can do it in constructor so `this.version` already contain range?
We can create two variables in classI think we should have two variables in class:
`this.version` - semver.Range
`this.rawVersion` - is raw input

File: src/distributors/base-installer.ts [hunk 38]
```diff
@@ -0,0 +1,121 @@
...
+        let foundJava = this.findInToolcache(range);
+
+        if(!foundJava) {
+            // download Java if it is not found locally
```
Comment (maxim-lobanov) [inline src/distributors/base-installer.ts:47]: Probably these comments can be removed

File: src/distributors/base-installer.ts [hunk 39]
```diff
@@ -0,0 +1,121 @@
...
+            javaVersion: getVersionFromToolcachePath(toolPath),
+            javaPath: toolPath
+        };
+    }
+
+    protected setJavaDefault(toolPath: string) {
+        core.exportVariable('JAVA_HOME', toolPath);
+        core.addPath(path.join(toolPath, 'bin'));
+        core.setOutput('path', toolPath);
+        core.setOutput('version', this.version);
```
Comment (maxim-lobanov) [inline src/distributors/base-installer.ts:77]: I think we should set resolved version here from `foundJava`

File: src/distributors/base-installer.ts [hunk 40]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    protected getJavaVersionsPath(): string {
```
Comment (maxim-lobanov) [inline src/distributors/base-installer.ts:80]: should we remove this function?

File: src/distributors/adoptopenjdk-installer.ts [hunk 41]
```diff
@@ -0,0 +1,79 @@
...
+import { JavaInstallerResults, IJavaRelease, JavaBase, JavaInstallerOptions } from "./base-installer";
+import { IRelease, IReleaseVersion } from './adoptopenjdk-models'
+
+export class AdoptOpenJdkDistributor extends JavaBase {
+    
+    constructor(initOptions: JavaInstallerOptions) {
+        super("AdoptOpenJDK", initOptions);
+    }
+
+    private async getAvailableMajor(range: semver.Range) {
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:18]: let's name this function as `resolveMajorVersion` and move to bottom

File: src/distributors/zulu-installer.ts [hunk 42]
```diff
@@ -0,0 +1,73 @@
...
+        toolPath = await tc.cacheDir(archivePath, `Java_${this.distributor}_${this.javaPackage}`, javaRelease.resolvedVersion, this.arch);
+
+        return { javaPath: toolPath, javaVersion: javaRelease.resolvedVersion };
+    }
+
+    protected async resolveVersion(range: semver.Range): Promise<IJavaRelease> {
+        const resolvedFullVersion = await this.getAvailableVersion(range);
+
+        const availableZuluReleaseUrl = `https://api.azul.com/zulu/download/community/v1.0/bundles/latest/?ext=${this.extension}&os=${this.platform}&arch=${this.arch}&hw_bitness=64&jdk_version=${resolvedFullVersion}&bundle_type=${this.javaPackage}`;
+        const availableZuluRelease = (await this.http.getJson<IZuluDetailed>(availableZuluReleaseUrl)).result;
```
Comment (maxim-lobanov) [inline src/distributors/zulu-installer.ts:45]: availableVersionsList

File: src/distributors/zulu-installer.ts [hunk 43]
```diff
@@ -0,0 +1,73 @@
...
+        const archivePath = path.join(extractedJavaPath, archiveName);
+        toolPath = await tc.cacheDir(archivePath, `Java_${this.distributor}_${this.javaPackage}`, javaRelease.resolvedVersion, this.arch);
+
+        return { javaPath: toolPath, javaVersion: javaRelease.resolvedVersion };
+    }
+
+    protected async resolveVersion(range: semver.Range): Promise<IJavaRelease> {
+        const resolvedFullVersion = await this.getAvailableVersion(range);
+
+        const availableZuluReleaseUrl = `https://api.azul.com/zulu/download/community/v1.0/bundles/latest/?ext=${this.extension}&os=${this.platform}&arch=${this.arch}&hw_bitness=64&jdk_version=${resolvedFullVersion}&bundle_type=${this.javaPackage}`;
```
Comment (maxim-lobanov) [inline src/distributors/zulu-installer.ts:44]: availableVersionsUrl

File: src/distributors/zulu-installer.ts [hunk 44]
```diff
@@ -0,0 +1,73 @@
...
+
+        return {link: availableZuluRelease.url, resolvedVersion: resolvedFullVersion};
+    }
+
+    private async getAvailableVersion(range: semver.Range): Promise<string> {
+        const availableVersionsUrl = `https://api.azul.com/zulu/download/community/v1.0/bundles/?ext=${this.extension}&os=${this.platform}&arch=${this.arch}&hw_bitness=64`;
+
+        const availableVersionsList = (await this.http.getJson<Array<IZulu>>(availableVersionsUrl)).result;
+
+        if(!availableVersionsList || availableVersionsList.length === 0) {
```
Comment (maxim-lobanov) [inline src/distributors/zulu-installer.ts:59]: nitpick: style around `if `

File: src/distributors/zulu-installer.ts [hunk 45]
```diff
@@ -0,0 +1,73 @@
...
+            throw new Error(`No Zulu java versions were not found for arch ${this.arch}, extenstion ${this.extension}, platform ${this.platform}`);
+        }
+
+        const zuluVersions = availableVersionsList.map(item => semver.coerce(item.jdk_version.join('.'))?? "");
```
Comment (maxim-lobanov) [inline src/distributors/zulu-installer.ts:63]: nitpick: style `item.jdk_version.join('.')) ?? ""`
Also should we filter null values?

File: src/distributors/zulu-installer.ts [hunk 46]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        const archiveName = fs.readdirSync(extractedJavaPath)[0];
+        const archivePath = path.join(extractedJavaPath, archiveName);
+        toolPath = await tc.cacheDir(archivePath, `Java_${this.distributor}_${this.javaPackage}`, javaRelease.resolvedVersion, this.arch);
+
+        return { javaPath: toolPath, javaVersion: javaRelease.resolvedVersion };
+    }
+
+    protected async resolveVersion(range: semver.Range): Promise<IJavaRelease> {
```
Comment (maxim-lobanov) [inline src/distributors/zulu-installer.ts:41]: may be rename `range` to `version` here and other places?

File: src/distributors/base-installer.ts [hunk 47]
```diff
@@ -0,0 +1,121 @@
...
+            allowRetries: true,
+            maxRetries: 3
+          });
+          this.version = this.normalizeVersion(initOptions.version);
+          this.arch = initOptions.arch;
+          this.javaPackage  = initOptions.javaPackage;
+    }
+
+    protected abstract downloadTool(javaRelease: IJavaRelease): Promise<JavaInstallerResults>;
+    protected abstract resolveVersion(range: semver.Range): Promise<IJavaRelease>;
```
Comment (maxim-lobanov) [inline src/distributors/base-installer.ts:40]: let's name this function as `findPackageForDownload`

File: src/distributors/adoptopenjdk-installer.ts (hunk 31 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 33 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 34 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/base-installer.ts (hunk 36 above)
Comment (dmitry-shibanov) [reply src/distributors/base-installer.ts]: Done.

File: src/distributors/base-installer.ts (hunk 38 above)
Comment (dmitry-shibanov) [reply src/distributors/base-installer.ts]: Done.

File: src/distributors/base-installer.ts (hunk 39 above)
Comment (dmitry-shibanov) [reply src/distributors/base-installer.ts]: Done.

File: src/distributors/base-installer.ts (hunk 40 above)
Comment (dmitry-shibanov) [reply src/distributors/base-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 41 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/zulu-installer.ts (hunk 44 above)
Comment (dmitry-shibanov) [reply src/distributors/zulu-installer.ts]: Done.

File: src/distributors/zulu-installer.ts (hunk 46 above)
Comment (dmitry-shibanov) [reply src/distributors/zulu-installer.ts]: Done.

File: src/distributors/base-installer.ts (hunk 47 above)
Comment (dmitry-shibanov) [reply src/distributors/base-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts [hunk 48]
```diff
@@ -0,0 +1,77 @@
...
+import { JavaInstallerResults, IJavaRelease, JavaBase, JavaInstallerOptions } from "./base-installer";
+import { IRelease, IReleaseVersion } from './adoptopenjdk-models'
+
+export class AdoptOpenJDKDistributor extends JavaBase {
+    
+    constructor(initOptions: JavaInstallerOptions) {
+        super("AdoptOpenJDK", initOptions);
+    }
+
+    private async resolveMajorVersion(range: semver.Range) {
```
Comment (maxim-lobanov) [inline src/distributors/adoptopenjdk-installer.ts:18]: could you please sort class methods the following way:
- public
- protected
- private

Please do this in adopt and azul classes

Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

File: src/distributors/zulu-installer.ts (hunk 45 above)
Comment (dmitry-shibanov) [reply src/distributors/zulu-installer.ts]: Done.

File: src/distributors/base-installer.ts (hunk 37 above)
Comment (dmitry-shibanov) [reply src/distributors/base-installer.ts]: Done.

File: src/distributors/base-installer.ts (hunk 35 above)
Comment (dmitry-shibanov) [reply src/distributors/base-installer.ts]: Done.

File: src/distributors/adoptopenjdk-installer.ts (hunk 32 above)
Comment (dmitry-shibanov) [reply src/distributors/adoptopenjdk-installer.ts]: Done.

**Pull request closed without merging**